skimage/exposure/_adapthist.py

```diff
@@ -109,106 +121,92 @@ def _clahe(image, ntiles_x, ntiles_y, clip_limit, nbins=128):
     minimum and maximum value as the input image. A clip limit smaller than 1
     results in standard (non-contrast limited) AHE.
     """
-    ntiles_x = min(ntiles_x, MAX_REG_X)
-    ntiles_y = min(ntiles_y, MAX_REG_Y)
 
     if clip_limit == 1.0:
         return image  # is OK, immediately returns original image.
 
-    h_inner = image.shape[0] - image.shape[0] % ntiles_y
-    w_inner = image.shape[1] - image.shape[1] % ntiles_x
-
-    # make the tile size divisible by 2
-    h_inner -= h_inner % (2 * ntiles_y)
-    w_inner -= w_inner % (2 * ntiles_x)
+    nr = int(np.ceil(image.shape[0] / kernel_size[0]))
+    nc = int(np.ceil(image.shape[1] / kernel_size[1]))
 
-    orig_shape = image.shape
-    width = w_inner // ntiles_x  # Actual size of contextual regions
-    height = h_inner // ntiles_y
+    row_step = int(np.floor(image.shape[0] / nr))
+    col_step = int(np.floor(image.shape[1] / nc))
 
-    if h_inner != image.shape[0]:
-        ntiles_y += 1
-    if w_inner != image.shape[1]:
-        ntiles_x += 1
-    if h_inner != image.shape[1] or w_inner != image.shape[0]:
-        h_pad = height * ntiles_y - image.shape[0]
-        w_pad = width * ntiles_x - image.shape[1]
-        image = pad(image, ((0, h_pad), (0, w_pad)), mode='reflect')
-        h_inner, w_inner = image.shape
+    img_view = view_as_windows(image, kernel_size, (row_step, col_step))
 
     bin_size = 1 + NR_OF_GREY // nbins
     lut = np.arange(NR_OF_GREY)
     lut //= bin_size
-    img_blocks = view_as_blocks(image, (height, width))
 
-    map_array = np.zeros((ntiles_y, ntiles_x, nbins), dtype=int)
-    n_pixels = width * height
-
-    if clip_limit > 0.0:  # Calculate actual cliplimit
-        clip_limit = int(clip_limit * (width * height) / nbins)
-        if clip_limit < 1:
-            clip_limit = 1
-    else:
-        clip_limit = NR_OF_GREY  # Large value, do not clip (AHE)
+    map_array = np.zeros((nr, nc, nbins), dtype=int)
 
     # Calculate greylevel mappings for each contextual region
-    for y in range(ntiles_y):
-        for x in range(ntiles_x):
-            sub_img = img_blocks[y, x]
+    for r in range(nr):
+        for c in range(nc):
+            if r < (nr - 1) and c < (nc - 1):
+                sub_img = img_view[r, c]
+            else:
+                sub_img = image[r * row_step: (r + 1) * row_step,
+                                c * col_step: (c + 1) * col_step]
```

Comment:
Although I like `view_as_windows` in principle, this expression is valid both in internal windows and on the edges, right? If so, I think we should just remove the windows view.